File: Backend-Nodejs/src/services/auth.service.js
// Importar módulo de encriptación
const crypto = require("crypto");
// Importar repositorio de usuarios
const userRepository = require("../repositories/user.repository");

// Función para encriptar contraseñas con MD5
function hashPassword(password) {
    return crypto.createHash("md5").update(password).digest("hex");
}

// Función para validar formato de correo electrónico
function validarCorreo(correo) {
    const regex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return regex.test(correo);
}



// Función principal para registrar un usuario
exports.register = async (data, file) => {

    // Extraer datos del objeto recibido
    const { correo, nombre_completo, password, confirm_password } = data;

    // Verificar que todos los campos estén presentes
    if (!correo || !nombre_completo || !password || !confirm_password) {
        throw new Error("Todos los campos son obligatorios");
    }

    // Validar que el correo tenga formato correcto
    if (!validarCorreo(correo)) {
        throw new Error("Formato de correo inválido");
    }

    // Verificar que las contraseñas coincidan
    if (password !== confirm_password) {
        throw new Error("Las contraseñas no coinciden");
    }

    // Buscar si el correo ya existe en la base de datos
    const existing = await userRepository.findByEmail(correo);
    if (existing) {
        throw new Error("El correo ya está registrado");
    }

    // Encriptar la contraseña
    const hashedPassword = hashPassword(password);

    // Crear ruta de la foto si se proporcionó un archivo
    const fotoKey = file ? `Fotos_Perfil/${file.originalname}` : null;

    // Guardar el usuario en la base de datos
    await userRepository.create(
        correo,
        nombre_completo,
        hashedPassword,
        fotoKey
    );

    // Retornar respuesta exitosa
    return {
        success: true,
        message: "Usuario registrado correctamente"
    };
};

// Función principal para iniciar sesión
exports.login = async (data) => {

    // Extraer datos del objeto recibido
    const { correo, password } = data;

    // Verificar que correo y contraseña estén presentes
    if (!correo || !password) {
        throw new Error("Correo y contraseña son obligatorios");
    }

    // Validar que el correo tenga formato correcto
    if (!validarCorreo(correo)) {
        throw new Error("Formato de correo inválido");
    }

    // Encriptar la contraseña ingresada
    const hashedPassword = hashPassword(password);

    // Buscar el usuario por correo en la base de datos
    const user = await userRepository.findByEmail(correo);

    // Verificar si el usuario existe
    if (!user) {
        throw new Error("Usuario no encontrado");
    }

    // Verificar si la contraseña coincide
    if (user.password !== hashedPassword) {
        throw new Error("Contraseña incorrecta");
    }

    // Retornar respuesta exitosa con datos del usuario
    return {
        success: true,
        message: "Login exitoso",
        user: {
            id_usuario: user.id_usuario,
            correo: user.correo,
            nombre_completo: user.nombre_completo,
            foto_perfil: user.foto_perfil
        }
    };
};

File: Backend-python/app/services/auth_service.py
# Importar módulo de hash para encriptación
import hashlib
# Importar módulo de expresiones regulares para validaciones
import re
# Importar funciones del repositorio de usuarios
from app.repositories.user_repository import find_by_email, create_user

# Función para encriptar contraseñas con MD5
def hash_password(password):
    return hashlib.md5(password.encode()).hexdigest()

# Función para validar formato de correo electrónico
def validar_correo(correo):
    return re.match(r"^[^\s@]+@[^\s@]+\.[^\s@]+$", correo)

# Función principal para registrar un usuario
def register_user(data, file):

    # Extraer datos del objeto recibido
    correo = data.get("correo")
    nombre = data.get("nombre_completo")
    password = data.get("password")
    confirm = data.get("confirm_password")

    # Verificar que todos los campos estén presentes
    if not correo or not nombre or not password or not confirm:
        raise Exception("Todos los campos son obligatorios")

    # Validar que el correo tenga formato correcto
    if not validar_correo(correo):
        raise Exception("Formato de correo inválido")

    # Verificar que las contraseñas coincidan
    if password != confirm:
        raise Exception("Las contraseñas no coinciden")

    # Buscar si el correo ya existe en la base de datos
    if find_by_email(correo):
        raise Exception("El correo ya está registrado")

    # Encriptar la contraseña
    hashed = hash_password(password)

    # Crear ruta de la foto si se proporcionó un archivo
    foto_key = f"Fotos_Perfil/{file.filename}" if file else None

    # Guardar el usuario en la base de datos
    create_user(correo, nombre, hashed, foto_key)

    # Retornar respuesta exitosa
    return {
        "success": True,
        "message": "Usuario registrado correctamente"
    }


# Función principal para iniciar sesión
def login_user(data):

    correo = data.get("correo")
    password = data.get("password")

    if not correo or not password:
        raise Exception("Correo y contraseña son obligatorios")

    if not validar_correo(correo):
        raise Exception("Formato de correo inválido")

    hashed_password = hash_password(password)

    user = find_by_email(correo)

    if not user:
        raise Exception("Usuario no encontrado")

    # En tu tabla:
    # id_usuario (0)
    # correo (1)
    # nombre_completo (2)
    # password (3)
    # foto_perfil (4)

    if user[3] != hashed_password:
        raise Exception("Contraseña incorrecta")

    return {
        "success": True,
        "message": "Login exitoso",
        "user": {
            "id_usuario": user[0],
            "correo": user[1],
            "nombre_completo": user[2],
            "foto_perfil": user[4]
        }
    }

